File: app/src/main/java/blagodarie/rating/ui/user/operations/OperationsFragment.java
package blagodarie.rating.ui.user.operations;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import org.jetbrains.annotations.NotNull;

import java.util.UUID;

import blagodarie.rating.AppExecutors;
import blagodarie.rating.R;
import blagodarie.rating.databinding.OperationsFragmentBinding;
import blagodarie.rating.model.entities.OperationType;
import blagodarie.rating.operations.OperationToAnyTextManager;
import blagodarie.rating.operations.OperationToUserManager;
import blagodarie.rating.repository.AsyncServerRepository;
import blagodarie.rating.server.BadAuthorizationTokenException;
import blagodarie.rating.ui.AccountProvider;
import blagodarie.rating.ui.OperationsFragmentArgs;

public final class OperationsFragment
        extends Fragment {

    public interface UserActionListener {
        void onThanksClick ();
    }

    private static final String TAG = OperationsFragment.class.getSimpleName();

    private OperationsViewModel mViewModel;

    private OperationsFragmentBinding mBinding;

    private OperationsAdapter mOperationsAdapter;

    private Account mAccount;

    private UUID mUserId;

    private UUID mAnyTextId;

    @NonNull
    private final AsyncServerRepository mAsyncRepository = new AsyncServerRepository(AppExecutors.getInstance().networkIO(), AppExecutors.getInstance().mainThread());

    @NonNull
    private final UserActionListener mUserActionListener = new UserActionListener() {
        @Override
        public void onThanksClick () {
            attemptToAddOperation(OperationType.THANKS, mUserId);
        }
    };

    @NotNull
    @Override
    public View onCreateView (
            @NonNull final LayoutInflater inflater,
            @Nullable final ViewGroup container,
            @Nullable final Bundle savedInstanceState
    ) {
        Log.d(TAG, "onCreateView");
        initBinding(inflater, container);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated (
            @NonNull final View view,
            @Nullable final Bundle savedInstanceState
    ) {
        Log.d(TAG, "onViewCreated");
        super.onViewCreated(view, savedInstanceState);

        final OperationsFragmentArgs args = OperationsFragmentArgs.fromBundle(requireArguments());

        mUserId = args.getUserId();
        mAnyTextId = args.getAnyTextId();
    }

    @Override
    public void onActivityCreated (@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        initViewModel();
        initOperationsAdapter();
        setupBinding();
    }

    @Override
    public void onResume () {
        Log.d(TAG, "onResume");
        super.onResume();
        refreshOperations();
    }

    @Override
    public void onDestroy () {
        Log.d(TAG, "onDestroy");
        super.onDestroy();
        mBinding = null;
    }

    private void initOperationsAdapter () {
        mOperationsAdapter = new OperationsAdapter();
    }

    private void initBinding (
            @NonNull final LayoutInflater inflater,
            @Nullable final ViewGroup container
    ) {
        Log.d(TAG, "initBinding");
        mBinding = OperationsFragmentBinding.inflate(inflater, container, false);
    }

    private void initViewModel () {
        mViewModel = new ViewModelProvider(requireActivity()).get(OperationsViewModel.class);
        mViewModel.isHaveAccount().set(mAccount != null);
        mViewModel.isOwnProfile().set(mAccount != null && mUserId != null && mAccount.name.equals(mUserId.toString()));
    }

    private void setupBinding () {
        /*mBinding.setViewModel(mViewModel);
        mBinding.setUserActionsListener(mUserActionListener);
        mBinding.rvOperations.setLayoutManager(new LinearLayoutManager(requireContext()));
        mBinding.rvOperations.setAdapter(mOperationsAdapter);
        mBinding.srlRefreshProfileInfo.setOnRefreshListener(this::refreshOperations);*/
    }

    private void refreshOperations () {
        mViewModel.getDownloadInProgress().set(true);
        Log.d(TAG, "refreshOperations");
        mViewModel.setOperations(
                mUserId != null ?
                        mAsyncRepository.getLiveDataPagedListFromDataSource(new UserOperationsDataSource.UserOperationsDataSourceFactory(mUserId)) :
                        mAsyncRepository.getLiveDataPagedListFromDataSource(new AnyTextOperationsDataSource.AnyTextOperationsDataSourceFactory(mAnyTextId))
        );
        mViewModel.getOperations().observe(requireActivity(), operations -> {
            mViewModel.getDownloadInProgress().set(false);
            mOperationsAdapter.submitList(operations);
        });
    }

    private void attemptToAddOperation (
            @NonNull final OperationType operationType,
            @Nullable final UUID userIdTo
    ) {
        Log.d(TAG, "onAddOperation");
        if (mAccount != null) {
            AccountProvider.getAuthToken(requireActivity(), mAccount, authToken -> {
                if (authToken != null) {
                    mAsyncRepository.setAuthToken(authToken);
                    if (userIdTo != null) {
                        new OperationToUserManager().
                                createOperationToUser(
                                        requireActivity(),
                                        UUID.fromString(mAccount.name),
                                        userIdTo,
                                        operationType,
                                        mAsyncRepository,
                                        () -> {
                                            Toast.makeText(requireContext(), R.string.info_msg_saved, Toast.LENGTH_LONG).show();
                                            refreshOperations();
                                        },
                                        throwable -> {
                                            Log.e(TAG, Log.getStackTraceString(throwable));
                                            if (throwable instanceof BadAuthorizationTokenException) {
                                                AccountManager.get(requireContext()).invalidateAuthToken(getString(R.string.account_type), authToken);
                                                attemptToAddOperation(operationType, userIdTo);
                                            } else {
                                                Toast.makeText(requireContext(), R.string.err_msg_not_saved, Toast.LENGTH_LONG).show();
                                            }
                                        });
                    } else {
                        new OperationToAnyTextManager().
                                createOperationToAnyText(
                                        requireActivity(),
                                        UUID.fromString(mAccount.name),
                                        mAnyTextId,
                                        operationType,
                                        "",
                                        mAsyncRepository,
                                        () -> {
                                            Toast.makeText(requireContext(), R.string.info_msg_saved, Toast.LENGTH_LONG).show();
                                            refreshOperations();
                                        },
                                        throwable -> {
                                            Log.e(TAG, Log.getStackTraceString(throwable));
                                            if (throwable instanceof BadAuthorizationTokenException) {
                                                AccountManager.get(requireContext()).invalidateAuthToken(getString(R.string.account_type), authToken);
                                                attemptToAddOperation(operationType, userIdTo);
                                            } else {
                                                Toast.makeText(requireContext(), R.string.err_msg_not_saved, Toast.LENGTH_LONG).show();
                                            }
                                        });
                    }
                } else {
                    Toast.makeText(requireContext(), R.string.info_msg_need_log_in, Toast.LENGTH_LONG).show();
                }
            });
        } else {
            AccountProvider.createAccount(
                    requireActivity(),
                    account -> {
                        if (account != null) {
                            mAccount = account;
                            mViewModel.isHaveAccount().set(true);
                            mViewModel.isOwnProfile().set(mAccount.name.equals(mUserId.toString()));
                            if (!mViewModel.isOwnProfile().get()) {
                                attemptToAddOperation(operationType, userIdTo);
                            } else {
                                Toast.makeText(requireContext(), R.string.info_msg_cant_add_operation_to_own_profile, Toast.LENGTH_LONG).show();
                            }
                        }
                    }
            );
        }
    }

}

File: app/src/main/java/blagodarie/rating/ui/operations/OperationsFragment.kt
package blagodarie.rating.ui.operations

import android.accounts.Account
import android.accounts.AccountManager
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.paging.PagedList
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import blagodarie.rating.AppExecutors
import blagodarie.rating.R
import blagodarie.rating.commands.CreateOperationToAnyTextCommand
import blagodarie.rating.commands.CreateOperationToUserCommand
import blagodarie.rating.databinding.OperationsFragmentBinding
import blagodarie.rating.model.IDisplayOperation
import blagodarie.rating.model.entities.OperationType
import blagodarie.rating.repository.AsyncServerRepository
import blagodarie.rating.server.BadAuthorizationTokenException
import blagodarie.rating.ui.AccountProvider
import blagodarie.rating.ui.AccountSource
import java.util.*

class OperationsFragment : Fragment() {

    interface UserActionListener {
        fun onThanksClick()
    }

    companion object {
        private val TAG = OperationsFragment::class.java.simpleName
    }

    private lateinit var mViewModel: OperationsViewModel

    private lateinit var mBinding: OperationsFragmentBinding

    private lateinit var mOperationsAdapter: OperationsAdapter

    private var mUserId: UUID? = null

    private var mAnyTextId: UUID? = null

    private val mAsyncRepository = AsyncServerRepository(AppExecutors.getInstance().networkIO(), AppExecutors.getInstance().mainThread())

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View {
        Log.d(TAG, "onCreateView")
        initBinding(inflater, container)
        return mBinding.root
    }

    override fun onViewCreated(
            view: View,
            savedInstanceState: Bundle?
    ) {
        Log.d(TAG, "onViewCreated")
        super.onViewCreated(view, savedInstanceState)
        val args = OperationsFragmentArgs.fromBundle(requireArguments())
        mUserId = args.userId
        mAnyTextId = args.anyTextId
        if (mUserId == null && mAnyTextId == null) {
            throw IllegalArgumentException("No userId and no anyTextId")
        }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        Log.d(TAG, "onActivityCreated")
        super.onActivityCreated(savedInstanceState)
        initViewModel()
        initOperationsAdapter()
        setupBinding()
    }

    override fun onResume() {
        Log.d(TAG, "onResume")
        super.onResume()
        refreshOperations()
        AccountSource.getAccount(
                requireActivity(),
                false
        ) {
            mViewModel.isOwn.set(it != null && it.name == mUserId.toString())
        }
    }

    private fun initOperationsAdapter() {
        Log.d(TAG, "initOperationsAdapter")
        mOperationsAdapter = OperationsAdapter(mViewModel.isOwn) {
            attemptToAddThanks(it, null)
        }
    }

    private fun initBinding(
            inflater: LayoutInflater,
            container: ViewGroup?
    ) {
        Log.d(TAG, "initBinding")
        mBinding = OperationsFragmentBinding.inflate(inflater, container, false)
    }

    private fun initViewModel() {
        Log.d(TAG, "initViewModel")
        mViewModel = ViewModelProvider(requireActivity()).get(OperationsViewModel::class.java)
    }

    private fun setupBinding() {
        Log.d(TAG, "setupBinding")
        mBinding.viewModel = mViewModel
        mBinding.list.recyclerView.adapter = mOperationsAdapter
        mBinding.refreshListener = SwipeRefreshLayout.OnRefreshListener {
            refreshOperations()
        }
        mBinding.userActionListener = object : UserActionListener {
            override fun onThanksClick() {
                attemptToAddThanks(mUserId, mAnyTextId)
            }
        }
    }

    private fun refreshOperations() {
        Log.d(TAG, "refreshOperations")
        mViewModel.downloadInProgress.set(true)
        if (mUserId != null) {
            mViewModel.operations = mAsyncRepository.getLiveDataPagedListFromDataSource(UserOperationsDataSource.UserOperationsDataSourceFactory(mUserId!!))
        } else if (mAnyTextId != null) {
            mViewModel.operations = mAsyncRepository.getLiveDataPagedListFromDataSource(AnyTextOperationsDataSource.AnyTextOperationsDataSourceFactory(mAnyTextId!!))
        }
        mViewModel.operations?.observe(requireActivity()) { pagedList: PagedList<IDisplayOperation>? ->
            mViewModel.isEmpty.set(pagedList?.isEmpty() ?: true)
            mOperationsAdapter.submitList(pagedList)
            mViewModel.downloadInProgress.set(false)
        }
    }

    private fun attemptToAddThanks(
            userIdTo: UUID?,
            anyTextIdTo: UUID?
    ) {
        Log.d(TAG, "onAddOperation")

        AccountSource.getAccount(
                requireActivity(),
                true
        ) { account: Account? ->
            if (account != null) {
                AccountProvider.getAuthToken(
                        requireActivity(),
                        account
                ) { authToken: String? ->
                    mAsyncRepository.setAuthToken(authToken)
                    if (userIdTo != null) {
                        CreateOperationToUserCommand(
                                requireActivity(),
                                UUID.fromString(account.name),
                                userIdTo,
                                OperationType.THANKS,
                                mAsyncRepository,
                                {
                                    Toast.makeText(requireContext(), R.string.info_msg_saved, Toast.LENGTH_LONG).show()
                                    refreshOperations()
                                },
                                { throwable: Throwable? ->
                                    Log.e(TAG, Log.getStackTraceString(throwable))
                                    if (throwable is BadAuthorizationTokenException) {
                                        AccountManager.get(requireContext()).invalidateAuthToken(getString(R.string.account_type), authToken)
                                        attemptToAddThanks(userIdTo, null)
                                    } else {
                                        Toast.makeText(requireContext(), R.string.err_msg_not_saved, Toast.LENGTH_LONG).show()
                                    }
                                }
                        ).execute()
                    } else if (anyTextIdTo != null) {
                        CreateOperationToAnyTextCommand(
                                requireActivity(),
                                UUID.fromString(account.name),
                                anyTextIdTo,
                                OperationType.THANKS,
                                "",
                                mAsyncRepository,
                                {
                                    Toast.makeText(requireContext(), R.string.info_msg_saved, Toast.LENGTH_LONG).show()
                                    refreshOperations()
                                },
                                { throwable: Throwable? ->
                                    Log.e(TAG, Log.getStackTraceString(throwable))
                                    if (throwable is BadAuthorizationTokenException) {
                                        AccountManager.get(requireContext()).invalidateAuthToken(getString(R.string.account_type), authToken)
                                        attemptToAddThanks(null, anyTextIdTo)
                                    } else {
                                        Toast.makeText(requireContext(), R.string.err_msg_not_saved, Toast.LENGTH_LONG).show()
                                    }
                                }).execute()
                    }
                }
            }
        }
    }
}